sound/soc/codecs/rt711-sdca.c

```diff
@@ -0,0 +1,1315 @@
+		RT711_DAC_DC_CALI_CTL1, RT711_DAC_DC_CALI_TRIGGER,
+		RT711_DAC_DC_CALI_TRIGGER);
+
+	/* wait for calibration process */
+	rt711_sdca_index_read(rt711, RT711_VENDOR_CALI,
+		RT711_DAC_DC_CALI_CTL1, &val);
+
+	for (loop = 0; loop < 500 &&
+		(val & RT711_DAC_DC_CALI_TRIGGER); loop++) {
+		usleep_range(10000, 11000);
+		ret = rt711_sdca_index_read(rt711, RT711_VENDOR_CALI,
+			RT711_DAC_DC_CALI_CTL1, &val);
+		if (ret < 0)
+			goto _cali_fail_;
+	}
+	if (loop == 500) {
+		dev_err(dev, "%s, calibration time-out!\n", __func__);
+		ret = -ETIMEDOUT;
+	}
+
+	/* release HP-JD and trigger FSM */
+	rt711_sdca_index_write(rt711, RT711_VENDOR_REG,
+		RT711_DIGITAL_MISC_CTRL4, 0x201b);
+
+_cali_fail_:
+	mutex_unlock(&rt711->calibrate_mutex);
+	dev_dbg(dev, "%s calibration complete, ret=%d\n", __func__, ret);
+	return ret;
+}
+
+static unsigned int rt711_sdca_button_detect(struct rt711_priv *rt711)
+{
+	unsigned int btn_type = 0, offset, idx, val, owner;
+	int ret;
+	unsigned char buf[3];
+
+	/* get current UMP message owner */
+	ret = regmap_read(rt711->regmap,
+		SDW_SDCA_CTL(FUN_HID, ENT_HID01, CTL_HIDTX_CURRENT_OWNER, 0),
+		&owner);
+	if (ret < 0)
+		return 0;
+
+	/* if owner is device then there is no button event from device */
+	if (owner == 1)
+		return 0;
+
+	/* read UMP message offset */
+	ret = regmap_read(rt711->regmap,
+		SDW_SDCA_CTL(FUN_HID, ENT_HID01, CTL_HIDTX_MESSAGE_OFFSET, 0),
+		&offset);
+	if (ret < 0)
+		goto _end_btn_det_;
+
+	for (idx = 0; idx < sizeof(buf); idx++) {
+		ret = regmap_read(rt711->regmap,
+			RT711_BUF_ADDR_HID1 + offset + idx, &val);
+		if (ret < 0)
+			goto _end_btn_det_;
+		buf[idx] = val & 0xff;
+	}
+
+	if (buf[0] == 0x11) {
+		switch (buf[1] & 0xf0) {
+		case 0x10:
+			btn_type |= SND_JACK_BTN_0;
+			break;
+		case 0x20:
+			btn_type |= SND_JACK_BTN_1;
+			break;
+		case 0x40:
+			btn_type |= SND_JACK_BTN_2;
+			break;
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+		switch (buf[2]) {
+		case 0x01:
+		case 0x10:
+			btn_type |= SND_JACK_BTN_0;
+			break;
+		case 0x02:
+		case 0x20:
+			btn_type |= SND_JACK_BTN_1;
+			break;
+		case 0x04:
+		case 0x40:
+			btn_type |= SND_JACK_BTN_2;
+			break;
+		case 0x08:
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+	}
+
+_end_btn_det_:
+	/* Host is owner, so set back to device */
+	if (owner == 0)
+		/* set owener to device */
+		regmap_write(rt711->regmap,
+			SDW_SDCA_CTL(FUN_HID, ENT_HID01, CTL_HIDTX_SET_OWNER_TO_DEVICE, 0),
+			0x01);
+
+	return btn_type;
+}
+
+static int rt711_sdca_headset_detect(struct rt711_priv *rt711)
+{
+	unsigned int det_mode;
+	int ret;
+
+	/* get detected_mode */
+	ret = regmap_read(rt711->regmap,
+		SDW_SDCA_CTL(FUN_JACK_CODEC, ENT_GE49, CTL_DETECTED_MODE, 0),
+		&det_mode);
+	if (ret < 0)
+		goto io_error;
+
+	switch (det_mode) {
+	case 0x00:
+		rt711->jack_type = 0;
+		break;
+	case 0x03:
+		rt711->jack_type = SND_JACK_HEADPHONE;
+		break;
+	case 0x05:
+		rt711->jack_type = SND_JACK_HEADSET;
```

Comment:
in SDCA the detected mode can be 'unknown' - e.g. some OEMs may want to let a user select headset or headphone, and the user can confirm with selected mode.

Reply:
> in SDCA the detected mode can be 'unknown' - e.g. some OEMs may want to let a user select headset or headphone, and the user can confirm with selected mode.

Do you mean that some kind of feature like UAJ? There is a pop-menu that a user could select headphone or headset, right?
If yes, the pop-menu is controlled by libgnome-volume-control.
Please check the function below.
https://github.com/GNOME/libgnome-volume-control/blob/e2be83ee4a47da9c4c4fbf302a63f04b8d5683b9/gvc-mixer-control.c#L2345
It needs a certain Jack control name, therefore, do we need another Jack control to report when the detected mode is 'unknown'?

Reply:
it's my understanding that there are already two controls in the GE (section 7.2.10.4 of SDCA v.06-r04), one for Detected Mode (RO) and Selected Mode (RW). The device can mirror Detected Mode and Selected Mode, but software has the ability to override any default. In case the detected mode is 'unknown" then indeed the selected mode needs to be selected by the user with a UI pop-up.

Reply:
OK, I understood. However, there are some application cases for GE in the RT711 SDCA.
I prefer using HW CBJ detection (one kind of auto-update mode) in the vendor-specific driver. 
In this mode, the detected_mode could report HP/Headset/Unplug, not including the unknown.

Reply:
I need to check if this is the expected behavior with SDCA, stay tuned.